File: C++/Find First and Last Position of Element in Sorted Array.cpp
//Name: Nurshat Fateh Ali

//Username: nurshatfateh

//Approach: Simple solution using modified version of binary search

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int>v;
        int s=0;
        int e=nums.size()-1;
        int m;
        int f=-1,l=-1;
        while(s<=e)
        {
            
            m=(s+e)/2;
            if(target==nums[m])
            {
              f=m;
              e=m-1;  
                
            
            }
            else if(target>nums[m])
            {
                s=m+1;
            }
            else if(target<nums[m])
            {
                e=m-1;
            }
     
            
        }
        s=0;
       e=nums.size()-1;
          while(s<=e)
        {
            
            m=(s+e)/2;
            if(target==nums[m])
            {
              l=m;
                s=m+1;
                
                
            
            }
            else if(target>nums[m])
            {
                s=m+1;
            }
            else if(target<nums[m])
            {
                e=m-1;
            }
     
            
        }
        
            v.push_back(f);
            v.push_back(l);
        return v;
    }
};

File: Java/Find_First_and_Last_Position_of_Element_in_Sorted_Array.java
class Solution {
    /*
        Rule 1
        - Given array is sorted in non-decreasing order
        
        Rule 2
        - Limit time complexity O(log n).
        
        So I can use Binary Search Algorithm.
    */
    public int[] searchRange(int[] nums, int target) {
        // initialize arr[0] -> maximum integer, arr[1] -> minimum integer
        int[] arr = {100001, -10};
        int s = 0;
        int e = nums.length - 1;
        
        // find minimum index in nums with Binary Search alogorithm
        while (s <= e) {
            
            int mid = (s + e) / 2;
            
            if(nums[mid] > target) {
                e = mid - 1;
            }
            else if(nums[mid] <= target) {
                if(nums[mid] == target) {
                    arr[0] = Math.min(arr[0], mid);
                    arr[1] = Math.max(arr[1], mid);
                }
                s = mid + 1;
            }
        }
        
        s = 0;
        e = nums.length - 1;
        
        // find maximum index in nums with Binary Search alogorithm
        while(s <= e) {
            int mid = (s + e) / 2;
            
            if(nums[mid] >= target) {
                if(nums[mid] == target) {
                    arr[0] = Math.min(arr[0], mid);
                    arr[1] = Math.max(arr[1], mid);
                }
                e = mid - 1;
            }
            else if(nums[mid] < target) {
                s = mid + 1;
            }
        }
        
        // if arr data is initial data, set -1.
        if(arr[0] == 100001) arr[0] = -1;
        if(arr[1] == -10) arr[1] = -1;
        
        return arr;
    }
}